Feature/shown posts

DMendybaeva opened the pull request.

DMendybaeva committed:
render active post information in modal window:

-add modal window in html
-add uiState
-add post id atribute in post btn

DMendybaeva committed:
render shownPosts with 'fw-normal' class

aRezn1kov commented on src/app.js:
```diff
@@ -77,6 +82,13 @@ const runApp = (t) => {
       });
   });
+
+  elements.modal.addEventListener('show.bs.modal', (e) => {
```
деструктуризация с кастомным именем
`{ relatedTarget: { dataset : { id: btnId } } }` - здесь `btnId` - кастомное имя

aRezn1kov commented on src/view/renderModalWindow.js:
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable no-param-reassign */
```
уберем в названии файла `Window`, оставим `renderModal`

reznikovAndrey commented on src/app.js:
```diff
@@ -77,6 +82,13 @@ const runApp = (t) => {
+    const btn = e.relatedTarget;
+    const btnId = btn.dataset.id;
+    const activePost = watchedState.posts.find(({ id }) => id === btnId);
+    watchedState.uiState.shownPosts = [activePost, ...watchedState.uiState.shownPosts];
```
лучше добавляй в конец, так будет быстрее с точки зрения оптимизации. 
Когда мы добавляем элемент в начало массива - нам приходится проходиться по всему массиву и сдвигать каждый элемент на 1 индекс вправо.

reznikovAndrey commented on src/view/renderModalWindow.js:
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable no-param-reassign */
+export default (value, { modal }) => {
+  const activePost = value[0];
```
здесь будем брать последний элемент. 
используй https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Array/at  - это новый метод у массивов, используй его с индексом -1

aRezn1kov replied:
Айдишник активного поста берешься из стейта. 
Дальше используешь метод 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get

aRezn1kov commented on src/view/renderPosts.js:
```diff
@@ -25,9 +25,12 @@ export default (value, { postsContainer }, t) => {
         'border-end-0',
       );
+
+      const isPostShown = shownPosts.map(({ id }) => id).includes(post.id);
```
- переименуй в `wasPostShown`
- `post.id` сделай как `postId`, для этого сделай деструктуризацию на 18 строке `forEach(({ id: postId, title }))`
- для проверки используй https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/has

reznikovAndrey commented on src/app.js:
```diff
@@ -23,6 +23,9 @@ const runApp = (t) => {
     processError: null, // сеть парсинг
     timerId: null,
+    uiState: {
+      shownPosts: [],
```
будем использовать вместо массива `Set`, чтобы избежать дублирования

reznikovAndrey replied:
Давай использовать `Map` вместо массива, где в качестве ключа мы будем использовать `id` поста, а в качестве значения - сам пост целиком. 

Это позволит избежать дублирования и добавит в оптимизации

aRezn1kov commented on src/app.js:
```diff
@@ -23,6 +23,9 @@ const runApp = (t) => {
     processError: null, // сеть парсинг
     timerId: null,
+    uiState: {
+      shownPosts: [],
```
добавим в `uiState` еще одно свойство: `activePostId`

DMendybaeva replied:
Добавила activePost и заменила shownPosts на shownPostsId

aRezn1kov commented on src/app.js:
```diff
@@ -23,6 +23,9 @@ const runApp = (t) => {
     processError: null, // сеть парсинг
     timerId: null,
+    uiState: {
+      shownPosts: [],
```
В качестве `shownPosts` будем использовать `Map`, где в качестве ключа будет айдишник поста, а в качестве значения - весь пост
https://learn.javascript.ru/map-set#map

DMendybaeva replied:
Использовала вместо Map, Set

aRezn1kov commented on src/app.js:
```diff
@@ -77,6 +82,13 @@ const runApp = (t) => {
+    const btn = e.relatedTarget;
+    const btnId = btn.dataset.id;
+    const activePost = watchedState.posts.find(({ id }) => id === btnId);
+    watchedState.uiState.shownPosts = [activePost, ...watchedState.uiState.shownPosts];
```
Здесь будем обновлять `activePostId` и потом добавлять в нашу коллекцию пост
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/set

DMendybaeva replied:
Сделала shownPostsId и activePost

aRezn1kov commented on src/view/view.js:
```diff
@@ -32,6 +33,9 @@ export default (state, elements, t) => onChange(state, (path, value) => {
     case 'timerId':
       break;
+    case 'uiState.shownPosts':
+      renderModalWindow(value, elements);
```
в функцию прокидывай `state`

DMendybaeva committed:
fix state, add hideShownPosts function

-add activePost state
-replace shownPosts to shownPostsId
-add hideShownPosts function
-remove hiding activePosts in renderModalWindow

DMendybaeva committed:
rename function, make destructuring

-rename renderModalWindow
-destructuring

reznikovAndrey commented on src/app.js:
```diff
@@ -23,6 +23,10 @@ const runApp = (t) => {
     processError: null, // сеть парсинг
     timerId: null,
+    uiState: {
+      shownPostsId: new Set(),
```
`shownPosts` делаем массивом

reznikovAndrey commented on src/app.js:
```diff
@@ -23,6 +23,10 @@ const runApp = (t) => {
     timerId: null,
+    uiState: {
+      shownPostsId: new Set(),
+      activePost: null,
```
удаляем

reznikovAndrey commented on src/app.js:
```diff
@@ -32,6 +36,8 @@ const runApp = (t) => {
     feedsContainer: document.querySelector('.feeds'),
     postsContainer: document.querySelector('.posts'),
     submitButton: document.querySelector('button[type="submit"]'),
+    postsButtons: document.querySelectorAll('button[data-bs-toggle="modal"]'),
```
это не нужно, можно удалить

aRezn1kov replied:
это нигде не используется, значит можем убрать

reznikovAndrey commented on src/app.js:
```diff
@@ -77,6 +83,13 @@ const runApp = (t) => {
+  elements.modal.addEventListener('show.bs.modal', ({ relatedTarget: { dataset: { id: btnId } } }) => {
+    const activePost = watchedState.posts.find(({ id }) => id === btnId);
+
+    watchedState.uiState.activePost = activePost;
```
не нужно

DMendybaeva committed:
fix hidePosts rendering during link click

aRezn1kov commented on src/app.js:
```diff
@@ -77,6 +83,13 @@ const runApp = (t) => {
   });
+
+  elements.postsContainer.addEventListener('click', ({ target }) => {
+    const activePost = watchedState.posts.find(({ id }) => id === target.dataset.id);
```
Если мы не нашли активный пост, то получим `undefined`
это нужно предусмотреть, чтобы не сыпались ошибки в консоль

aRezn1kov commented on src/view/hideShownPosts.js:
```diff
@@ -0,0 +1,5 @@
+export default (state, elements) => {
```
смени название файла на `renderShownPost`

aRezn1kov commented on src/view/renderModal.js:
```diff
@@ -0,0 +1,15 @@
+export default ({
+  title: actPostTitle,
+  description: actPostDescription,
+  linkPost: actLinkPost,
```
здесь не нужны кастомные переменные, подойдут дефолтные

aRezn1kov commented on src/view/view.js:
```diff
@@ -32,6 +34,12 @@ export default (state, elements, t) => onChange(state, (path, value) => {
+      break;
     default:
       throw new Error(`Unknown path: ${path}`);
   }
```
давай переименуем этот файл в `index.js`

DMendybaeva committed:
fix naming

DMendybaeva committed:
fix error in activePostId finding
